crypto/stack/stack.c

```diff
@@ -118,32 +118,57 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
     return (NULL);
 }
 
-int OPENSSL_sk_insert(OPENSSL_STACK *st, const void *data, int loc)
+int OPENSSL_sk_reserve(OPENSSL_STACK *st, int n)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
+    size_t target, num_alloc, sz;
+
+    if (st == NULL || st->num < 0)
         return 0;
-    }
 
-    if (st->num_alloc <= (size_t)(st->num + 1)) {
-        size_t doub_num_alloc = st->num_alloc * 2;
-        const char **tmpdata;
+    if (n <= 0)
+        return 1;
+
+    /* Sanity checks */
+    if (n > INT_MAX - st->num
+        || (size_t)n > SIZE_MAX / sizeof(char *) - st->num)
+        return 0;
+
+    target = st->num + n;
+    for (num_alloc = st->num_alloc; num_alloc <= target; num_alloc = sz) {
+        sz = num_alloc * 2;
 
         /* Overflow checks */
-        if (doub_num_alloc < st->num_alloc)
+        if (sz < num_alloc)
             return 0;
 
         /* Avoid overflow due to multiplication by sizeof(char *) */
-        if (doub_num_alloc > SIZE_MAX / sizeof(char *))
+        if (sz > SIZE_MAX / sizeof(char *))
```

Comment:
Actually this is not the only boundary. It makes no sense allocate more than INT_MAX[+1] elements, because maximum index in int.

On related note one can wonder if doubling is actually right strategy. Well, it gets kind of "schizophrenic" here. Doubling is probably right in start, then it should rather switch to linear, but you are probably unlikely to even need that many entries. And since amount of entries is not large, then checks for SIZE_MAX feels excessive...

Reply:
By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.
I've limited the size to INT_MAX (pushed separately but will be squashed).

As for the doubling, I wondered about this too.  I don't know the right strategy.  Do we have data as to the size stacks can grow to inside OpenSSL?  It is a public API, so we can't know for sure.

Reply:
As for the size increase, the best reference I've found so far is _Why Are Vectors Efficient?_ by Andrew Koenig from the _Journal of Object-Orientated Programming_ volume 11, number 5, published in 1998.  I've not been able to locate the actual article but the result is presented in [this comp.land.c++moderated article](https://groups.google.com/d/msg/comp.lang.c++.moderated/asH_VojWKJw/Lo51JEmLVboJ).

The conclusion is that growth factors greater than:
![Φ = 1.618...](https://latex.codecogs.com/gif.latex?%5CPhi&space;=&space;%5Cfrac{1&space;&plus;&space;sqrt(5)}{2}&space;%5Capprox&space;1.618)
aren't great.  I've found other discussions that suggest linear growth is never great, so I'm going to try exponential growth with a factor of 1.5.

Reply:
> By my reading of the standard, it looks like SIZE_MAX can be less than INT_MAX.

But in all practical cases it's vice versa and what I'm talking about is rather relation between `SIZE_MAX/sizeof(void *)` vs. `INT_MAX` than `SIZE_MAX` and `INT_MAX`.

> I've limited the size to INT_MAX (pushed separately but will be squashed).

Formally speaking the limit should be `(size_t)INT_MAX+1`. Indeed, last permitted index being INT_MAX means that there are INT_MAX+1 elements allocated. Here is another relevant question. Imagine going back to doubling for a moment, and imagine you wanted to reserve INT_MAX-13 elements. With INT_MAX as cap it would fail, because algorithm would search for closest 2^n larger than requested amount but it will fail check for INT_MAX. For wrong reason. Well, one case of course argue that INT_MAX-13 should never happen and we might as well let it slide. But it does make code counter-intuitive. Check for INT_MAX+1 would address this. But now consider proposed 1.5 growth factor... I don't want to calculate problematic values, what I'm rather trying to ask is if `return 0` as outcome of this test is right? Maybe setting setting  to INT_MAX+1 is more adequate? And by "right" and "adequate" I rather mean rather "least counter-intuitive" than "practical"...

Reply:
It seems that I'm wrong about INT_MAX+1 thing. stack_st.num is amount of elements, not last index, and it's integer. This means that last possible index is INT_MAX-1, not INT_MAX, as I suggested. But the question about whether or not corner cases similar to INT_MAX-13 should be treated as valid (again, not from practical viewpoint, but formal) remains...

Reply:
I think I've addressed the INT_MAX-13 concern, I limit growth to INT_MAX now.

I left the check against SIZE_MAX/sizeof(char *).  There is an earlier check for n+st->num against INT_MAX that prevents that being too large and the loop is checking for INT_MAX.  These prevent num_alloc being greater than INT_MAX but it is still possible that size_t cannot handle that many pointers (e.g. when sizeof(int) == sizeof(size_t)).  Thus, I think the SIZE_MAX check is correct.